### PR TITLE
Merge gitlab to GitHub

### PR DESCRIPTION
**Pull request by nitishgoyal13**

### CODE REVIEW

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java [hunk 1]
```diff
@@ -56,9 +54,36 @@ public TimeWindow getWindow() {
     }
 
     @Override
-    @JsonIgnore
-    public boolean isFilterTemporal() {
-        return true;
+    public int hashCode() {
+        int result = getOperator().hashCode();
```
Comment (nitishgoyal13) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java]: Needed to exclude timestamp field from hashcode for caching to work. Caching doesn't work if the same query gets executed in gap of less than 30s

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/DataStore.java [hunk 2]
```diff
@@ -36,4 +36,6 @@
     Document get(final Table table, final String id);
 
     List<Document> getAll(final Table table, final List<String> ids);
+
+    void updateTable(final Table table);
```
Comment (nitishgoyal13) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/datastore/DataStore.java]: This is added to update ttl in HBase table

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchUtils.java [hunk 3]
```diff
@@ -138,6 +139,10 @@ private static XContentBuilder getDocumentMapping() throws IOException {
                 .startObject()
                 .field("enabled", false)
                 .endObject()
+                .field("_all")
```
Comment (nitishgoyal13) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchUtils.java]: Do we need add this in upstream? We don't store documents in ES

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/reroute/ClusterRerouteConfig.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.flipkart.foxtrot.core.reroute;
+
+import com.flipkart.foxtrot.core.jobs.BaseJobConfig;
+import lombok.Data;
+
+/***
+ Created by mudit.g on Feb, 2019
+ ***/
+@Data
```
Comment (nitishgoyal13) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/reroute/ClusterRerouteConfig.java]: This job is added to re route shards every day in the morning to balance the cluster

File: foxtrot-server/src/main/java/com/flipkart/foxtrot/server/resources/ClusterHealthResource.java [hunk 5]
```diff
@@ -64,7 +75,17 @@ public NodesStatsResponse getNodeStat() throws ExecutionException, InterruptedEx
     @Timed
     @Path("indicesstats")
     @ApiOperation("getIndicesStat")
-    public IndicesStatsResponse getIndicesStat() throws ExecutionException, InterruptedException {
-        return queryStore.getIndicesStats();
```
Comment (nitishgoyal13) [inline foxtrot-server/src/main/java/com/flipkart/foxtrot/server/resources/ClusterHealthResource.java]: This is changed to send additional information like  avg document size, number of columns on the cluster info page

File: foxtrot-server/src/main/java/com/flipkart/foxtrot/server/resources/DocumentResource.java [hunk 6]
```diff
@@ -169,33 +123,46 @@ private String preProcess(String table, Document document) {
...
-                Document document = iter.next();
-                if(document.getData()
-                        .has(EVENT_TYPE)) {
-                    String eventType = document.getData()
-                            .get(EVENT_TYPE)
-                            .asText();
-                    if(eventToBeIgnored(table, eventType)) {
-                        LOGGER.info("Skipped event for table: {} with eventType: {} Document: {}", table, eventType, document);
-                        iter.remove();
-                    }
```
Comment (nitishgoyal13) [inline foxtrot-server/src/main/java/com/flipkart/foxtrot/server/resources/DocumentResource.java]: Shall we add this in upstream ? This is added to send events to different tables? Splitting of tables based on eventTypes

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java [hunk 7]
```diff
@@ -56,9 +54,36 @@ public TimeWindow getWindow() {
...
+        result = 31 * result + getField().hashCode();
+        if(!getField().equals("_timestamp")) {
+            result = result * 21 + (getCurrentTime() == 0
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java]: In the span of 5 lines, getCurrentTime() is called 3 times

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java [hunk 8]
```diff
@@ -56,9 +54,36 @@ public TimeWindow getWindow() {
...
+        }
+        else if(!(o instanceof LastFilter)) {
+            return false;
+        }
+
+        LastFilter that = (LastFilter) o;
+
+        return getField().equals(that.getField()) && getOperator().equals(that.getOperator()) &&
+                getDuration().equals(that.getDuration()) && getRoundingMode().equals(that.getRoundingMode()) &&
+                getCurrentTime() == that.getCurrentTime();
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java]: This will never match even if the difference is 1 ms

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java [hunk 9]
```diff
@@ -56,9 +54,36 @@ public TimeWindow getWindow() {
     }
 
     @Override
-    @JsonIgnore
-    public boolean isFilterTemporal() {
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/datetime/LastFilter.java]: Why is this removed?

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/numeric/BetweenFilter.java [hunk 10]
```diff
@@ -80,6 +76,44 @@ public boolean isFilterTemporal() {
...
+        if(!getField().equals("_timestamp")) {
+            result = result * 21 + (getFrom() == null
+                                    ? 43
+                                    : getFrom().hashCode());
+            result = result * 23 + (getTo() == null
+                                    ? 29
+                                    : getTo().hashCode());
+        }
+        else {
+            result = result * 21 + Long.valueOf(getFrom().longValue() / (long) 30000).hashCode();
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/numeric/BetweenFilter.java]: It is not correct to do this for anything other than `last` filter, this behaviour is not correct. We can have an additional flag  which can govern if cached results are acceptable. But for critical cases, people expect correct results and this will definitely not wield one.

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/numeric/BetweenFilter.java [hunk 11]
```diff
@@ -80,6 +76,44 @@ public boolean isFilterTemporal() {
...
+        }
+        else if(!(o instanceof BetweenFilter)) {
+            return false;
+        }
+
+        BetweenFilter that = (BetweenFilter) o;
+
+        return getField().equals(that.getField()) && getOperator().equals(that.getOperator()) &&
+                isFilterTemporal() == that.isFilterTemporal() && getFrom().equals(that.getFrom()) &&
+                getTo().equals(that.getTo());
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/numeric/BetweenFilter.java]: Logic looks shaky. Please check. Use the cache acceptable flag here to govern the behaviour

File: foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/numeric/NumericBinaryFilter.java [hunk 12]
```diff
@@ -61,4 +59,37 @@ public boolean isFilterTemporal() {
...
+    public int hashCode() {
+        int result = getOperator().hashCode();
+        result = 31 * result + getField().hashCode();
+        if(!getField().equals("_timestamp")) {
+            result = result * 21 + (getValue() == null
+                                    ? 43
+                                    : getValue().hashCode());
+        }
+        else {
+            result = result * 21 + Long.valueOf(getValue().longValue() / (long) 30000).hashCode();
```
Comment (santanusinha) [inline foxtrot-common/src/main/java/com/flipkart/foxtrot/common/query/numeric/NumericBinaryFilter.java]: Check comments. It is not correct to do this indiscriminately. Create and abstract function to get timestamp hash code.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java [hunk 13]
```diff
@@ -59,15 +59,18 @@
     private final ElasticsearchConnection elasticsearchConnection;
     private final CacheManager cacheManager;
     private final ObjectMapper objectMapper;
+    private final ElasticsearchTuningConfig elasticsearchTuningConfig;
 
     public AnalyticsLoader(TableMetadataManager tableMetadataManager, DataStore dataStore, QueryStore queryStore,
-                           ElasticsearchConnection elasticsearchConnection, CacheManager cacheManager, ObjectMapper objectMapper) {
+            ElasticsearchConnection elasticsearchConnection, CacheManager cacheManager, ObjectMapper objectMapper,
```
Comment (santanusinha) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/spi/AnalyticsLoader.java]: Move everything to one line each

File: foxtrot-server/src/main/resources/console/echo/fql/index.htm [hunk 14]
```diff
@@ -87,10 +87,16 @@ <h1 class="col-sm-4 heading">
...
+                    <li><a href="../cluster/index.htm">Cluster</a></li>
                     <li><a id="active-fql">FQL</a></li>
                     <li id="fql-dashboard">
                         <span id="create-new-dashboard">Create New Dashboard</span>
                     </li>
+                    <li>
+                        <form id="logout" method="post">
+                            <img id="logout-icon" class="logout" src="../img/logout.png" title="Logout" alt="Logout">
+                        </form>
+                    </li>
```
Comment (santanusinha) [inline foxtrot-server/src/main/resources/console/echo/fql/index.htm]: There is no login .. :D

File: foxtrot-server/src/main/resources/console/echo/index.htm [hunk 15]
```diff
@@ -63,9 +63,16 @@ <h1 class="col-sm-4 heading">
                 </select>
                 <ul id="left-menu">
                     <li><a href="browse-events.htm">Browse Events</a></li>
+                    <li><a href="cluster/index.htm">Cluster</a></li>
+                    <li><a href="fql/index.htm">FQL</a></li>
                     <li id="add-sections" data-target="#addDashboard" data-toggle="modal">
                         <span id="create-new-dashboard">Create New Dashboard</span>
                     </li>
+                    <li>
+                        <form id="logout" method="post">
```
Comment (santanusinha) [inline foxtrot-server/src/main/resources/console/echo/index.htm]: Remove from everywhere
